### PR TITLE
Improvements on spm parity:

### LINKED ISSUE #53: Feature Request: Support for sentencepiece unigram model
Any plans to add support for sentencepiece unigram model?

### PR DESCRIPTION
**Pull request by Narsil**
- Adding `encoding_optimized` version of the `encode` algorithm.
   - `Trie<char>` -> `Trie<u8>` Another improvement on speed.
   - We can now detect a lot more errors
   - Removes Lattice allocation.
   - Changes trie `common_prefix_search` to return an iterator to avoid
    allocation of the full results.
   - Adding 2 new normalizers that mimick spm defaut's behavior.
   - Improving *a lot* the parity check.
   - Adding `Precompiled` tokenizer that is 100% the `SentencePiece` normalizer and is able to be read from `python`.

Reformer/Pegasus | Flaubert | XLM | CTRL are BPE, not Unigram

| File | T5  | Albert | Camembert | MBart | XLMRoberta | Marian | XLNet | 
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | 
| big.txt |✔️99.98⁽³⁾ (3.6x)⁽²⁾ |✔️99.99 (3.6x) | ✔️99.85 (3.55x) | ✔️ 99.96 (3.5x) | ✔️99.96 (3.5x) | ✔️99.8 (3.86x) | ✔️99.99 (3.5x)
| wagahima.txt (japanese) | ✔️99.87 (2x) | ✔️99.87 (1.7x) |  ✔️100 (1.98x) | ✔️100 (3x) | ✔️100 (3x) | ✔️99.87 (2.36x) | ✔️100 (1.97x)
| wikitext-103 (raw, 500Mo) | ✔️99.99 (3.37x) | ✔️99.99 (3.14x) | ✔️99.99 (3.11x) | ❌⁽¹⁾ 99.99 (3.25x) | ❌⁽¹⁾ 99.99 (3.23x) | ❌⁽¹⁾⁽⁴⁾ 99.99 (3.37x) | ✔️ 99.99 (3.452x)
|xnli|✔️99.98 (3.17) | ✔️99.99 (3.0) | ✔️99.97 (3.12) |  ❌⁽⁵⁾ 99.98 (3.8) | ❌⁽⁵⁾ 99.98 (4.1) | ❌⁽⁴⁾⁽⁵⁾ 99.97 (3.1) | ✔️99.99 (3.0)

(1): There is one known flaw, it's the japanese tilde (～ vs ~). The standard unicode normalization nfkc changes ～ into ~, but `spm` removes that single rule from the `nfkc` standard normalization, we currently keep to the standard.
(2): The number in parenthesis `(3.6x)` indicates that we are slower than `spm`. This is due in part because of not yet optimized code, and in part because of the way we normalize and pretokenize which keeps tracks of all changes to be able to revert them which is more costly than `spm` lossy normalize part.
(3): Checkmark indicates that we are 100% compliant. The number `99.98` indicates the percentage of perfectly accurate tokenization. The remainder was checked and is equivalent to `spm`'s implementation or we are correcting a probable bug in `spm`. There are 2 kinds of imperfections:

- "AAA" can be equivalently tokenized as "AA" + "A" and "A" + "AA", we handle floats differently explaining the differences between `spm` and `tokenizers`.
- The second class is a probable bug where some words are tokenized differently by `spm` in the same sentence.
  For instance with `t5` tokenizer the sentence
```python
sp.encode('increasing to 35 ( km / h ) Lonaconing Lonaconing .')
# [3094, 12, 3097, 41, 2280, 3, 87, 3, 107, 3, 61, 301, 106, 9, 1018, 53, 1815, 11962, 106, 53, 3, 5]
```
"Lonaconing" is both split as` '▁L', 'on', 'a', 'con', 'ing'` and ` '▁Lo', 'nac', 'on', 'ing',`. In `tokenizers` we consistently tokenize as the first form. This bug only happens for long sentences, never for small ones, so the way we check that is that both `spm` and `tokenizers` encode the mistokenized part in the same way (meaning that for `spm` full sentence tokenization and simple word **differ**)

(4): Although this [model](https://s3.amazonaws.com/models.huggingface.co/bert/Helsinki-NLP/opus-mt-en-de/source.spm) claims to be trained with standard `nmt-nfkc`. It seems the unicode zero-width character "\u200d" is ignored by default but does not seem to be so in this model, the whole `nmt` part seems to be missing actually. It was either trained with an old version, or overloaded some rules. We will need to load the `precompiled_chars_map` to fully support this model.

(5): It seems `spm` does not handle nkfc for right to left languages correctly. In the string `'صوَّرت U2 الكثير من الأفلام.'`, unicode normalization seems that `\xd9\x91\xd9\x8e` needs to be changed to `xd9\x8e\xd9\x91`. We can check with:
```python
import unicodedata

string = b"\xd9\x91\xd9\x8e".decode('utf-8')
string == 'َّ'
nkfc_string = unicodedata.normalize("NFKC", string).encode('utf-8')
# b'\xd9\x8e\xd9\x91'
nfkc_string != string
```

Another example is with "ết" (`b'\xe1\xba\xbft'` vs `b'\xc3\xaa\xcc\x81'`)

Fixes #53 53

### CODE REVIEW

File: tokenizers/src/normalizers/precompiled.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        trie_blob.push(n.to_usize());
+    }
+    let normalized_blob = rest;
+    assert_eq!(rest.len() + trie_size as usize + 4, n);
+    Ok((normalized_blob, trie_blob))
+}
+
+impl Precompiled {
+    pub fn from(precompiled_charsmap: &[u8]) -> Result<Precompiled> {
+        let (normalized_blob, trie_blob) = parse(&precompiled_charsmap).unwrap();
```
Comment (n1t0) [inline tokenizers/src/normalizers/precompiled.rs:147]: We should probably avoid the `.unwrap()` here too, no?

**Review by n1t0 — COMMENTED** [review]
Very nice!

File: tokenizers/src/models/unigram/model.rs [hunk 2]
```diff
@@ -202,17 +206,127 @@ impl Unigram {
...
             self.cache.set(sentence.to_owned(), result.clone());
             result
         }
     }
-    fn encode_no_cache(&self, sentence: &str) -> Vec<String> {
-        // TODO optimized version
+
+    pub fn encode_optimized(&self, sentence: &str) -> Vec<String> {
         // https://github.com/google/sentencepiece/blob/d48247191a6d50e469ed1a4a36e877befffd1851/src/unigram_model.cc#L600
+        #[derive(Debug, Clone)]
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/models/unigram/model.rs:223]: I mimicked spm here, and this Struct is only used in this very function.

I could move them, but I felt that this file was not appropriate (as this struct is very local and not relevant to Unigram in general). Do you think `encode_optimised.rs` would be a better location for this code ? I am not sure.

NB: note to myself, this function probably should not be `pub`.

File: tokenizers/src/normalizers/mod.rs [hunk 3]
```diff
@@ -1,11 +1,16 @@
 pub mod bert;
+pub mod precompiled;
+// Separatae file because of long bytes.
+#[cfg(test)]
+mod precompiled_test;
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/mod.rs:5]: I am not sure I know how to get a valid binary string otherwise that still manages to get some tests off.

This one is `spm` `nmt_nfkc` (default) normalization rule, which still deviates from the standard (tilde, and some right to left inconsistencies with nfkc). That's a great way to test this as they are the edge cases that cause problems with the tests listed.

Maybe I could reduce the file size by having the binary be directy encoded and not be u8 ? We don't care about readability anyway.

File: tokenizers/src/normalizers/precompiled.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        state.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for Precompiled {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_map(PrecompiledVisitor)
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/precompiled.rs:39]: Neat. Didn't know how to do that.

File: tokenizers/src/normalizers/precompiled.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        // No, I don't know why Google did this.
+        // If you question this code, check this normalizer against
+        // XNLI database (all languages) with Unigram model against
+        // Mbart, XLMRoberta *AND* Marian. If you don't get 100% or
+        // break a single test.
+        // You don't pass.
+        normalized.get().graphemes(true).for_each(|grapheme| {
+            if grapheme.len() < 6 {
+                if let Some(norm) = self.transform(grapheme) {
+                    for (i, c) in norm.chars().enumerate() {
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/precompiled.rs:196]: Should they all be `-N` or only the first one ? I wasn't sure when trying this.

File: bindings/python/py_src/tokenizers/implementations/sentencepiece_model_pb2.py [hunk 6]
```diff
@@ -0,0 +1,1150 @@
+# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
[the remark this answers is not in the document]
Comment (Narsil) [reply bindings/python/py_src/tokenizers/implementations/sentencepiece_model_pb2.py]: Done.

File: bindings/python/src/models.rs [hunk 7]
```diff
@@ -263,21 +263,21 @@ pub struct PyUnigram {}
 #[pymethods]
 impl PyUnigram {
     #[new]
```
[the remark this answers is not in the document]
Comment (Narsil) [reply bindings/python/src/models.rs:265]: I don't think it's a good idea to have a default value, it's an index into the vocab list, if we get the default wrong, we will have a bogus Model.

File: bindings/python/src/normalizers.rs [hunk 8]
```diff
@@ -273,6 +279,36 @@ impl Normalizer for PyNormalizerWrapper {
...
+        Ok((PyNmt {}, Nmt.into()))
+    }
+}
+
+#[pyclass(extends=PyNormalizer, module = "tokenizers.normalizers", name=Precompiled)]
+pub struct PyPrecompiled {}
+#[pymethods]
+impl PyPrecompiled {
+    #[new]
+    fn new(precompiled_charsmap: Vec<u8>) -> PyResult<(Self, PyNormalizer)> {
```
[the remark this answers is not in the document]
Comment (Narsil) [reply bindings/python/src/normalizers.rs:297]: Done.

File: tokenizers/src/models/unigram/trie.rs [hunk 9]
```diff
@@ -30,23 +30,50 @@ impl<Label: Eq + Hash + Copy> Trie<Label> {
...
+
+impl<Label, T> Iterator for TrieIterator<'_, Label, T>
+where
+    Label: Eq + Hash + Copy,
+    T: Iterator<Item = Label>,
+{
+    type Item = Vec<Label>;
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.iterator.next() {
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/models/unigram/trie.rs:59]: Done.

File: tokenizers/src/models/unigram/trie.rs [hunk 10]
```diff
@@ -30,23 +30,50 @@ impl<Label: Eq + Hash + Copy> Trie<Label> {
...
+    prefix: Vec<Label>,
+    iterator: T,
+}
+
+impl<Label, T> Iterator for TrieIterator<'_, Label, T>
+where
+    Label: Eq + Hash + Copy,
+    T: Iterator<Item = Label>,
+{
+    type Item = Vec<Label>;
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/models/unigram/trie.rs:56]: Well I had a hard time doing it, and the solution I found seemed unusable : https://stackoverflow.com/questions/30422177/how-do-i-write-an-iterator-that-returns-references-to-itself It does not work because now the vec is not owned anymore by the iterator, so I can't push into it anymore, that would require mutable ref, so I can't hand non mutable ref anymore.

File: tokenizers/src/normalizers/precompiled.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (n1t0) [reply tokenizers/src/normalizers/precompiled.rs:196]: The second variable of the tuple (named `change` in `transform`) gives the number of changes brought by the associated `char`. So if it comes before `N` removed chars, you have to specify `-N`. It is supposed to be `1` for added characters, and `0` for unchanged/replaced ones (cf [transform docs](https://github.com/huggingface/tokenizers/blob/7b2caca/tokenizers/src/tokenizer/normalizer.rs#L593-L602)).

File: tokenizers/src/models/unigram/model.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (n1t0) [reply tokenizers/src/models/unigram/model.rs:223]: Nope, this is a method on `Unigram`, so it seems a great location for it. I just found that it makes it harder to read, but don't feel strongly about this. I wouldn't be surprised to see a struct used by only one function defined at the file level, as long as the visibility is set appropriately.

File: tokenizers/src/normalizers/precompiled.rs [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        let maybe_type_str = maybe_type.as_ref().map(|(k, v)| (k.as_str(), v));
+        match maybe_type_str {
+            Some(("precompiled_charsmap", value)) => Ok(Precompiled::from(&value)
+                .map_err(|_| Error::custom("Cannot read `precompiled` string"))?),
+            _ => Err(Error::custom("Expected precompiled value, got {:?}")),
+        }
+    }
+}
+
+named!(be<u32>, u32!(Endianness::Little));
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/precompiled.rs:63]: Done.

File: tokenizers/src/normalizers/precompiled.rs [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+            }
+        }
+        results
+    }
+}
+
+fn parse(precompiled_charsmap: &[u8]) -> IResult<&[u8], Array> {
+    let n = precompiled_charsmap.len();
+    let (mut rest, trie_size) = le_u32(precompiled_charsmap)?;
+    assert_eq!(trie_size % 4, 0);
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/precompiled.rs:131]: I removed the assert's nom should take care of that automatically in follow up calls. (Returning a custom error seems super hard in `nom`)

File: tokenizers/src/normalizers/precompiled.rs (hunk 1 above)
Comment (Narsil) [reply tokenizers/src/normalizers/precompiled.rs:147]: Done

File: tokenizers/src/tokenizer/normalizer.rs [hunk 13]
```diff
@@ -124,6 +124,21 @@ pub struct NormalizedString {
 }
 
 impl NormalizedString {
+    pub fn new(
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/tokenizer/normalizer.rs:127]: Hmm it shouldn't have to be public, I really just needed some constructor for my `precompiled_test` so I can test normalized, original and both alignments.

I switched this constructor to `pub(crate)` but I don't like it either. Do you think there's another way ?

File: tokenizers/src/tokenizer/mod.rs [hunk 14]
```diff
@@ -601,10 +601,16 @@ where
         offsets_type: OffsetType,
     ) -> Result<Encoding> {
         let encode = |is_pre_tokenized, subseq_idx, subseq| -> Result<Encoding> {
+            let now = std::time::Instant::now();
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/tokenizer/mod.rs:604]: Woops.

File: tokenizers/src/models/unigram/model.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/models/unigram/model.rs:223]: Ok, I'll leave it here if you don't feel strongly about it.

File: tokenizers/src/normalizers/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (n1t0) [reply tokenizers/src/normalizers/mod.rs:5]: We can maybe have it in a separate file, that we download specifically for the tests. We would need to extract the tests as integration, and download from the `Makefile`, but I prefer this solution than having this huge source code file.

File: tokenizers/src/models/unigram/trie.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (n1t0) [reply tokenizers/src/models/unigram/trie.rs:56]: Alright, we can't indeed return a reference on the content of the Iterator itself.

File: tokenizers/src/normalizers/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/mod.rs:5]: I don't think unit tests should rely on external files as much as possible.

I agree having this huge thing is not great, but I think downloading it as external is even worse. It make tests depend on an internet connexion, it makes logic of tests hard to understand, it can fail tests because of flaky connection, making tests flaky. It also hinder the test readability and so on.

I tried having this binary string in raw format within the source code (which bring the source file to ~233k) but it makes the file really unreadable with superduper long string, that because it's invalid utf8 messes with editors. 

The other option which I think is viable, is to split the Precompiled struct into an external crate that can tolerate this large file within it's own tests and we just implement the Normalizer trait in `tokenizers`. It's another way of making that downloadable, without actually being downloaded by the other crate.
It does add a dependency though :(

[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/mod.rs:5]: Building tests in release mode (for bench for instance) with this huge file does mean compilation is *especially* slow.

[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/mod.rs]: I extracted it as a separate crate. So I think this is correct.

File: tokenizers/src/normalizers/unicode.rs [hunk 15]
```diff
@@ -36,7 +36,98 @@ impl Normalizer for NFKC {
     }
 }
 
+fn do_nmt(normalized: &mut NormalizedString) {
+    // Ascii Control characters
+    normalized
+        .filter(|c| match c as u32 {
+            0x0001 => false,
```
[the remark this answers is not in the document]
Comment (Narsil) [reply tokenizers/src/normalizers/unicode.rs]: Done.

**Comment by loretoparisi** [comment]
@Narsil I'm not sure this will fix

```javascript
    let encoder = (tokenizer) => promisify(tokenizer.encode.bind(tokenizer))
    let decoder = (tokenizer) => promisify(tokenizer.decode.bind(tokenizer))

    const bertTokenizer = await SentencePieceBPETokenizer.fromOptions({  vocabFile: '/sentencepiece.bpe.model"  });
    
    encode = encoder(bertTokenizer);
    decode = decoder(bertTokenizer);

    encoded = await encode("Hello how are you?");
    console.log(encoded.getIds());
    console.log(encoded.getTokens());
    decoded = await decode(encoded.getIds(), true);
    console.log(decoded);
```

that outputs a `[]` array. Thank you!
